### DISCUSSION TITLE
Google Summer of Code Discussion

### DISCUSSION THREAD

**Discussion by tpike3** [mesa/mesa#2046] [General]
If you are interested in supporting Mesa during the Google Summer of Code, please ask questions or other inquiries here!

[GSoC Idea List](https://github.com/projectmesa/mesa/wiki/Projects-for-Google-Summer-of-Code-2024)

**Comment by apranav22** [comment]
Hey! Has any work been done on these ideas given in the list previously?

**Reply by rht** [reply]
The ones with the links do.

**Reply by Wei-cyber** [reply]
Hello. project ideas look very interesting. Should we work on the initial proposal right now or do we have to submit a contribution as mentioned in the contributor guide ?(https://github.com/projectmesa/mesa/blob/main/CONTRIBUTING.md)

**Reply by rht** [reply]
The application opens on March 18, which is ~a month from now, so you have plenty of time. Submitting small contributions is not required, but would help us in evaluating the potential contributors, given the limited number of slots.

**Reply by Wei-cyber** [reply]
Thanks

**Reply by Prithvij2004** [reply]
Hi there! I just wanted to say that I find the project ideas really interesting and appealing. I'm actually currently learning about RL algorithms myself, and I think it would be a great opportunity to contribute to the project.

**Reply by tpike3** [reply]
@Prithvij2004 Great to meet you and let us know if you have any questions

**Comment by Prithvij2004** [comment]
I have little knowledge about Reinforcement Learning, but I would like to
learn more about RL. Also I am new to open source contributing. I quite
didn't catch the RL problem mentioned in the tasks. It would be great if
you could tell me where I could get the resources.

**Reply by tpike3** [reply]
@Prithvij2004 Here are some libraries with tutorials you should find useful. Let us know if you have any more questions: 

- https://openrl-docs.readthedocs.io/en/latest/
- https://keras.io/examples/rl/   ;  https://keras-rl.readthedocs.io/en/latest/
- https://www.gymlibrary.dev/content/tutorials/

**Reply by Prithvij2004** [reply]
Hey, I have a doubt regarding DQN, where can I contact you? This is based on the code I have seen leading frameworks made. Can we discuss it here? Or should I mail you?

**Reply by tpike3** [reply]
Apologies just saw this you can email me at tpike3@gmu.edu or post here.

**Comment by madman024** [comment]
Hello everyone, i want to work on the network scheduler project ,to clarify i dont know a lot of things ,but i am willing to learn many things on the way, how can i get started on the project

**Reply by EwoutH** [reply]
Start looking into the current Mesa codebase, check the examples, built some models, and get a feeling about what the current limitations are around Networks and scheduling.

Also read #1912. It's an important open discussions we're currently having within Mesa.

**Reply by madman024** [reply]
yeah got it ,thanks a lot,if any need i willl reach out to you

**Comment by ankitmrj** [comment]
Greetings!

 


I am Ankit Srivastav , a first year student of Bachelor of Technology (Computer Science and Engineering) from MMMUT India.
I have experience in the field of MERN development. I am open source contributor and eager to contribute to esteemed organization.
I am thrilled to contribute and learn from idea Cacheable Mesa.
Please help way to contribute ,discuss and share ideas more.

Email: vaibhavsribhai@gmail.com
Ankit Srivastav

**Reply by tpike3** [reply]
Welcome @ankitmrj let us know if you have any questions

**Comment by manvendra5345** [comment]
Hi everyone!
I am Manvendra Singh Solanki from Cse Ai-Ml background and i am also interested in working in this project , I know python , java , HTML , CSS  . I can adapt very fast and have great learning capacity but these thing are common what i think i have different in me is that i want to learn from thoes who are better than me and catch up  to them and even beat them and always keep learning.

Email:- singhmanvendra2003@gmail.com

**Reply by tpike3** [reply]
Welcome @manvendra5345 let us know if you have any questions

**Comment by VibhuAg** [comment]
Hi all,
I'm interested in working on integrating RL libraries into Mesa. I currently work on RL research and have experience using libraries like PettingZoo for MARL. I'm looking forward to contributing to the Mesa community by bringing in my knowledge of RL and MARL!

**Reply by EwoutH** [reply]
Welcome! If you have any questions, let us know

**Comment by scramblerdoodle** [comment]
Hey!
I'm currently in my last year of studies in Applied Mathematics and, after having taken a course on Emergence in Complex Systems (in which we used ABMs to study their behaviour) during my exchange program, I became very interested in the subject!
I've been working as a Python developer for the past 6 years and I've been wanting to get into open source projects since then, and I would love to contribute to Mesa.

I'm interested in the Cacheable Mesa project idea, since I have some experience with storing states for mathematical models, and I had a few questions:
Is the main idea to follow up on the issues related to this topic, as well as continuing to implement the related forks (for example [mesa-interactive](https://github.com/Corvince/mesa-interactive) and [mesa-replay](https://github.com/Logende/mesa-replay))? Or is it more like a catch-all of the different ideas as examples on what can be done to implement cache on Mesa?
I'd love to discuss further any points of attention related to this project idea!

As a side note, I've been already messing around in building the project locally and running the test suite, and have been looking at the issues I could start contributing with. If there are any relevant contributions I can already get started on, do let me know!

**Comment by adamamer20** [comment]
Do you have a specific template for the proposal or following the guidelines [here](https://google.github.io/gsocguides/student/writing-a-proposal#elements-of-a-quality-proposal) will do?
Thanks!

**Reply by EwoutH** [reply]
There is no required template, following the guidelines is a good idea. But feel free to structure it as you see fit (form follows function).

**Comment by vmaciels** [comment]
Hi there!
I'm very interested in contributing to the scheduler project. I have a masters in network science, and although I have much to learn in the programming part, I think I can help a lot in including complex networks into the library.
About the inclusion of the scheduler as a part of a dynamic network, would the focus be on implementing it from scratch in the project itself, or using other libraries like networkx? And is the idea for it to be constrained to the spatial module, or maybe implementing a new topological module?

**Reply by EwoutH** [reply]
I think other libraries often have the preference above custom code we need to maintain.

If you can, join our dev call Tuesday. Think about use cases for such a scheduler beforehand. This feels like an idea worth discussing in person.

**Reply by EwoutH** [reply]
Going though example libraries is often a good way to get familiar to ABM use cases. However, be aware of survivor bias, things that are hard to implement might be less example for.

**Comment by Soyvor** [comment]
Hello everyone i want to contribute to the mesa Rl Project and be a part of it as i wanna be a part of a big community and also be a part of gsoc, I have experience with  Reinforced Learning and i have been working with various libraries under Rl and would love to be a part of it, i just need guidance on where to start from.

**Reply by EwoutH** [reply]
Welcome! For Mesa RL, some work that has been done:

- https://github.com/projectmesa/mesa/discussions/2070
- https://github.com/projectmesa/mesa/discussions/2075

**Comment by Jacksonxhx** [comment]
Hi, I'm Jackson Xie, currently studying computer science and math in NYU, I'm experienced in developing with python and have some knowledge about ml, nlp and application of them. I'm currently working in a Financial AI company as a software engineer. To be honest, i did not have the experience of developing in an open source project before, but that's also the reason i'm here. I'm responsible and really interested in this project, specifically the Mesa RL project. I will try this repo and play with it these days and hope to contribute to the project!

**Reply by EwoutH** [reply]
Welcome! If you have any question please let me know.

**Comment by Increshi** [comment]
Hello all, I am Rushi currently doing my Btech at the Indian Institute of Technology Tirupati. I have a good experience in Python and Cpp. Also, I have good knowledge of ML algorithms, DL, and Reinforcement learning. I have also done my internship at the Robert Bosch Center of Data Science and Artificial Intelligence, IIT Madras, where I worked on Gen AI. I am also familiar with Python Libraries Numpy, Matplotlib, pandas, and ML frameworks like Pytorch, Tensorflow, and Keras.I also have experience with Git and Github. I am very interested in the MesaRL Project as it explores RL which I like most.And I love playing with RL Algos like PPO, Q learning etc. Looking forward to working with you.

**Reply by EwoutH** [reply]
Welcome! If you have any questions please let us know.

**Comment by partizanos** [comment]
Hello, I am Dimitrios Proios, currently a 3rd-year PhD student in the Faculty of Medicine at the University of Geneva in Switzerland, with a focus on representation learning using heterogeneous temporal graph neural network models. I found the **Network Scheduler proposal** fascinating and aligned with my research interests.  I am proficient in Python and primarily use PyTorch for machine learning, though I have also worked with TensorFlow. Additionally, I have utilized JAX to teach reinforcement learning (Q-learning, bandit algorithms) at the Mediterranean Machine Learning Summer [School in 2024](https://github.com/M2Lschool/tutorials2023). My experience includes working at CERN and in financial institutions as a web developer, data engineer, and quantitative developer with 8 years of work experience in both research and industry ([LinkedIn](https://www.linkedin.com/in/dproios/), [Google Scholar](https://scholar.google.com/citations?user=ef3dKRMAAAAJ&hl=en&oi=ao)), it would be a pleasure to contribute to the project in the coming months through source contribution, as well as provide maintenance and support as an active member of the open-source community.

**Reply by EwoutH** [reply]
Welcome! If you have any questions please let us know.

**Comment by madman024** [comment]
hey @tpike3 @EwoutH @rht sry to tag u all ,but i was filling my gsoc application but now theres no submission form,ig theres still time as its  4:47 utc,what should i do.

**Reply by EwoutH** [reply]
Can you post a screenshot of what you do see? Are you properly logged in? If correct, you should have another 56 minutes.

**Reply by EwoutH** [reply]
Maybe try to see if you can reach GSoC support in some way.

**Reply by madman024** [reply]
@EwoutH hey i wasnt able to fill ,i filled the proposal form and then it reloded and the button and form to submit proposal were not there i dont know what was the issues .thanks for your msg.btw could u just go through my proposal and give me feedback,that would be helpful for me.

**Comment by Samarpan1122** [comment]
Hi everyone! 👋

I'm Samarpan Mohanty, a first-year Computer Engineering student at UNL and a Research Assistant in Dr. Bavarian's lab, where we use machine learning to advance solid-state battery research.

I have experience working with ML ,Python, C#, JavaScript, SQL, and Java for projects and research, and I also enjoy learning new languages just for fun.

I'm particularly interested in open-source contributions, and I aspire to be a GSoC contributor in 2025. Looking forward to learning more about Mesa and engaging with the community! 🚀

email - samarpanmohanty3347@gmail.com

**Comment by JJC022** [comment]
I am intrigued by the issues in the GSoC 2025 writeup. Is there any room for overlap between Mesa-LLM and the behavioral exploration? Additionally, what would the specific implementation of behaviors look like? Would it be an explicit MDP? Would there be any default behavioral frameworks present, and if yes how would they be contextualized to a user's custom model out of the box? 

It seems there is a lot of interest in integrating deepRL in this thread. Would the goals of the people interested in RL be able to serve the goals of the maintainers who introduced the behavioral framework project? In terms of use cases, is the goal to introduce reliable behavioral frameworks to the mesa codebase, or is the goal to provide a tool for exploring how behavior policies emerge in modeling frameworks?

I am a Cognitive Science student at UCSD and I have been working on a traffic modelling project using Mesa for the last 10 weeks. The next step for my research is to model different behavior policies for agents to reflect heterogenous decision making in traffic, so I will be doing some work in this direction anyways. I would really appreciate guidance / feedback from the Mesa maintainers in this endeavor.

**Reply by EwoutH** [reply]
These are excellent questions, and I'm happy to discuss them! I currently don't have much more ideas than I wrote down already at this point (check al the links in the idea) though.

One goal of the behavioral framework is to make it easier for Agents to self-optimise their behaviour, a.k.a. goal seeking / self preservation. Last year we had an RL project, I felt that was the right direction but fundamental building blocks in Mesa were still missing.

I would say dig in, and once you have some ideas I'm happy to discuss them!

**Reply by JJC022** [reply]
Thank you for your reply. I read a bit more on the writeup for the Behavioral Framework project and read some of the discussions / pull requests linked there. 

I think this is a really unique and fascinating approach to behavioral modeling - in particular BDI and Motivation-Drive theory are really different from what I have learned about in this area previously. I have some background in Sociology and currently do Ecology-related work, so I am really interested in the kinds of models that are being mentioned here. 

While there is clearly a lot of interest in RL from the people commenting here, I haven't seen anyone address the unique demands of implementing these things in Mesa. in line with your discussions / pull requests, I agree that  RL / Behavior in Mesa should be grounded in ContinuousStates and Tasks. In particular, I am interested in how states, actions, and rewards would be handled within Mesa. Your framework of States, Decisions, Actions, Goals and Events really interests me. However, while some have suggested integrating RLib or another library that runs off of Gym seems challenging. It seems impractical to use a gym .env to train agents in a Mesa space. Devs and Observations seem important to consider. My worry is that developing this idea without being in constant dialogue with all of the other features will hurt its success. For instance, the batch runner / mesa frames seem like they could be instrumental in parallelizing training here. 

Like someone else mentioned (sorry, not familiar with all of the Maintainers yet), I would be concerned about how all of this is going to be presented to the user in a easy to use way. It seems that Mesa's value is that it can provide a really quick and powerful way to build a model with lots of potential for building something bigger from there. 

I am also interested in the LLM-modeling project, but personally I am more interested in deploying more lightweight neural nets that have been trained in the model they will be deployed in. My biggest question is how does this project exist in relation to plans for bringing features out of the experimental folder and into the main library? 

I will take more time to dig in and clarify my ideas, but I wanted to get my thoughts out there despite them being disorganized.

**Reply by EwoutH** [reply]
@JJC022 my current perspective is that we need the behavioral framework first in Mesa, before trying to use LLM’s or RL makes any sense. I think the framework is an essential building block to be able to do the latter two.

However, it’s fully okay not to build the full thing at once. If you can fledge out a few of the components in GSoC while leaving some hooks to in the future build the others, that would already be very useful.

Looking forward to your future thoughts!

**Comment by invcble** [comment]
Hi everyone, Imon here! MSCS grad student from Drexel looking to contribute to Mesa's exciting ideas. For the past years I have been working on various end-to-end decision/RAG pipelines (powered by LLMs) for various industries (medical, GTM etc). I also have open-sourced a simple windows utility app which has over 4500 downloads in total. I aim to understand how to contribute to a large codebase and at the same time learn from the experts in the field.

The 2nd idea Mesa-LLM caught my attention. While I was unfamiliar with the Mesa framework in the beginning, I have started my research on it and from my initial understanding (I will continue to learn more) it looks a very cool tool to simulate "agents" in a controlled environment. I looked up some sample setups and understood how behaviors are currently hardcoded for agents and hence the introduction of Mesa-LLM.

I idea itself really excites me, cause I am very much familiar with the openai, ollama, llamacpp sdks and this is a great feature addon if successfully implemented. That being said, while I am learning more about this repo, if anyone who is reading this message, can direct me what essential things I have to take care first to be able to contribute and direct me to a place where I can show my initial approach to the problem and get feedback on it, I would be grateful.

**Reply by wang-boyu** [reply]
Thanks for your interest. Please feel free to share your work/thoughts here. There's also another discussion https://github.com/projectmesa/mesa/discussions/2679 to talk about mesa-llm.

**Comment by shockylove** [comment]
Hello everyone,

My name is Xiaoke Li, a Master's student in Computer Science at the University of Illinois Urbana-Champaign. I'm excited about the opportunity to contribute to the Mesa project, particularly in integrating reinforcement learning and advancing agent-based modeling capabilities.

## Relevant Experience

As a Research assistant under Prof. Lingming Zhang at UIUC, I've worked extensively with large language models, focusing on stabilizing deep reasoning through chain-of-thought constraints and supervision of deepseek- R1 intermediate steps. This experience has given me valuable insights into developing robust AI systems that can be applied to various domains, including agent-based modeling.

My internships at Alibaba Group, AGIBOT, and ByteDance have provided me with substantial software engineering experience. At Alibaba, I built a hybrid recommendation system that combined collaborative filtering with BERT-based content embeddings, directly addressing cold-start challenges—a problem that relates closely to improving user experience in complex systems like those modeled in Mesa.

## Technical Skills

- **Programming Languages**: Proficient in Python, Java, C/C++, and SQL
- **AI/ML Frameworks**: Extensive experience with PyTorch, TensorFlow, and Hugging Face Transformers
- **NLP Expertise**: Developed BERT-based pipelines for structured data extraction and optimized transformer models for deployment
- **Backend Systems**: Experience with Docker, Kubernetes, AWS, and CI/CD pipelines
- **Frontend Development**: Familiarity with AngularJS and Vue.js for creating user-friendly interfaces

## Initial Contribution Plan

To begin contributing, I propose addressing a small but meaningful issue: improving the documentation and examples for the existing state, action, and reward handling mechanisms in Mesa. This will help new users understand the current capabilities while laying the groundwork for more advanced RL integration.

I'll start by:
1. Reviewing the current documentation and identifying gaps or areas for clarification
2. Creating improved examples that demonstrate state, action, and reward implementation
3. Submitting a pull request with these documentation enhancements
4. Opening a discussion thread to gather feedback on this initial contribution and plan next steps for RL integration

## Questions for the Community

1. How does the Mesa community envision the integration of RL libraries like RLib or Gym? Are there specific design patterns or architectural considerations we should follow?
2. What are the community's thoughts on developing a standardized interface for different RL libraries within Mesa?
3. How can we ensure that new RL features remain compatible with existing Mesa functionality like the BatchRunner and DataCollector?
4. Are there any existing experimental features related to RL that might be candidates for migration to the main library?

I'm particularly interested in how we can make these complex modeling capabilities accessible to users while maintaining the power and flexibility of Mesa. I look forward to collaborating with you all on this exciting project!

Best regards,
Xiaoke Li

**Comment by gadhvirushiraj** [comment]
Hello all,

I've been sketching out some ideas for the structure of the mesa-llm project. After working with current models to understand how Mesa operates, I've also been experimenting with https://github.com/microsoft/tinytroupe over the past few days. Given the broad scope of mesa-llm, there are multiple potential directions we could take.

I had a couple of questions, to help me understand more and align towards the UX part of the project idea.

1. Most research on agentic simulations have focused on diff. architectures of memory management, how memory flows are structured for both the environment and the agent. Memory still remains a pain point in current agentic model research.  While reasoning, CoT, and similar functionalities can be implemented as API calls where users can modify the models and prompts. But for memory design, if provided with building blocks it would still require a lot of debugging while building a model-sim. TinyTroupe only provides episodic and semantic memory implementation, while you can make fast iteration you lack on experimenting with diff arch.

I'm still a bit uncertain about which approach to consider. @wang-boyu, could you share your thoughts on whether there's a particular subset of agentic modeling we are targeting with a project or plans for tackling memory design? For example, replicating this memory from work  https://arxiv.org/abs/2304.03442 would require orchestration of all diff types and won't be possible by supporting a single memory method.

2. Support for external libraries?   LangChain and similar libraries are widely used, which is a clear advantage if integrated. However, their frequent updates could pose maintenance challenges. How open is Mesa to external dependencies and to what extent should we abstract or integrate such lib’s solutions?

3. I've noticed that many agentic AI implementations lack a robust visualization component. While some research has experimented with sandbox environments, a comprehensive, integrated tool is still missing. Even tracking conversations in multi-agent systems is a challenge. Do you have any ideas in this area to implement?

Thanks

**Reply by EwoutH** [reply]
@wang-boyu @Sahil-Chhoker (and others!) curious of your thoughts on this!

For memory, also check out [Behavioral Framework #2538](https://github.com/projectmesa/mesa/discussions/2538). What I call "States" there is basically internal agent memory.

**Reply by wang-boyu** [reply]
Thanks for sharing the TinyTroupe repo. I didn't know that there's another framework from Microsoft on llm agents apart from AutoGen. To answer your questions:

> whether there's a particular subset of agentic modeling we are targeting with a project

No I can't think of any particular subset of models. For mesa-llm I would prefer it to be generic enough so that it can be used in various kinds of simulations (e.g., social simulations, task solving, and so on) with llm-based agents.

> or plans for tackling memory design

Since we're creating mesa-llm from scratch, I will leave this question to our applicants to answer in their proposals : ) Really looking forward to them.

> How open is Mesa to external dependencies and to what extent should we abstract or integrate such lib’s solutions?

So far Mesa has been fairly flexible about its external dependencies, so long as they are needed. Similar to the answer above, mesa-llm doesn't exist yet, so we are open to ideas about whether to use existing frameworks or not. That said, I would recommend to keep APIs fexible enough to be framework agnostic while designing them. So that later on if we decide to switch to using or dropping dependencies on frameworks, our users wouldn't be affected too much. External frameworks are viewed as "implementation details" hidden from users as much as possible.

> I've noticed that many agentic AI implementations lack a robust visualization component. ... Do you have any ideas in this area to implement?

Mesa uses Solara for model visulaiztion and it's possible to develop custom components for it, much like how Mesa creates custom visualizations for model spaces, gis spaces, measure plots, etc. Mesa-LLM should similarly use Solara for visualizations, with custom components if needed.

**Comment by suryanshgargbpgc** [comment]
Hi @tpike3, @EwoutH, and @adamamer20,

I have some questions regarding the proposal submission . Apart from the implementation of the idea, what key areas should I focus on to strengthen my proposal?  Additionally, how can I connect personally for a review of my proposal.

Looking forward to your insights.

**Reply by EwoutH** [reply]
Thanks for reaching out!

- Really think about your target audience and what value it provides in you proposal
- Think about how it fits in Mesa (how it interacts with other Mesa components)
- Target one small atomic thing and do that really well, don't try to jump too far into the future
- We don’t review full proposals. However, we do encourage to discuss specific ideas with us and each other, in the open, and get excited about it! :)
- Build something in Mesa! (it will help you with all above)

**Reply by suryanshgargbpgc** [reply]
Thanks for your reply, however can you provide clarification on following,
1. Is there any template which I can follow ?
2. In "about me" section should I add information about the projects which I have done and it necessarily have not much to do with the current project ?


> * We don’t review full proposals.
3. okay, but if my initial draft requires some upgrades, then how can I fix them, like I might end up overexplaining on a point which was not required.

**Reply by EwoutH** [reply]
No template except what Google provides. Form follows function. Don't overdo the about me section, let your work speak for itself (preferably here in Mesa).

We simply don't have the maintainer capacity to review full proposals, we're all volenteers and there's a lot of GSoC interest.

At 13:30 CET we have a call, you can ask questions of discuss ideas there after the main topics.

**Reply by suryanshgargbpgc** [reply]
> At 13:30 CET we have a call, you can ask questions of discuss ideas there after the main topics.
sure, thanks for your help

**Comment by soh-123** [comment]
Hi 

Before submitting my porposal for 'Frontend Upgrade', should I make a contribution first? additionally, is there a porposal template to follow?

**Reply by EwoutH** [reply]
Hi, thanks for reaching out!

The order does not matter. After the review deadline we will look both at the proposals and contributions up to that point.

However, your experience in contributing might help you write a better proposal. We don't have a template other than Google provides, form follows function.

**Comment by Bpriya42** [comment]
Hello everyone and @EwoutH @adamamer20 @tpike3 !

I'm Priya, a Master's student in Computer Science at UMass Amherst, specializing in Natural Language Processing. I recently came across MESA and was particularly excited about your work on Agentic LLMs, which aligns closely with my current research project on agentic AI systems.

I'm looking forward to engaging with the community and would love to be considered for GSoC 2025. Additionally, I wanted to ask if there are any specific technical requirements or expectations you'd like to see addressed in the proposal.

Thanks so much!
Priya Balakrishnan
[Website](https://bpriya42.github.io/ProfileWebsite/)

**Comment by suryanshgargbpgc** [comment]
hi @adamamer20 @EwoutH @tpike3 @quaquel, I am submitting proposal for mesa-frames upgrade, however length of the project is mentioned as 175/350 hrs, so what should I put there since there is a disclaimer not to change the length without organizations approval, based on the proposal I have prepared, it seems to be a large project.

**Comment by ShreyasN707** [comment]
Hi everyone!

I’m a second-year AIML student, and I recently started getting into open source. I’m also learning machine learning algorithms, and I found your projects really interesting to contribute to. I’m hoping to participate in GSoC as well, I’d like some guidance on how I can get involved and start contributing to this project.
